Fix missing drag image on FF, Safari

pleku opened the pull request:
When CSS transform has been applied, the drag image is missing (safari),
or gets offset (FF). Fixed by using custom drag image without transform,
and for Safari tries to cancel transforms on parent DOM tree. Does NOT fix #9408

Fixes #9261

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9409)
<!-- Reviewable:end -->

Commit:
Fix missing drag image on FF, Safari

When CSS transform has been applied, the drag image is missing (safari),
or gets offset (FF). Fixed by using custom drag image without transform,
and checking for transforms on parent DOM tree. Does NOT fix #9408

Fixes #9261

hesara commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java, line 246 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp4BvkJvCDk-ExpggY:-Kkp4BvkJvCDk-ExpggZ:b-t92vv) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java#L246)):*
> ```Java
>                     } else {
>                         // Separate data with new line character when multiple
>                         // rows
> ```

(could consolidate comment on two lines)

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java, line 475 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp45oK6Rz4e6Ya4YwO:-Kkp45oK6Rz4e6Ya4YwP:b-nis7ng) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java#L475)):*
> ```Java
>         });
>     }
> }
> ```

(EOL missing)

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 237 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp1267nYQhnAYPxujk:-Kkp1267nYQhnAYPxujl:b-srspmg) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L237)):*
> ```Java
>      * Fixes missing or offset drag image caused by using css transform:
>      * translate (or such) by using a cloned drag image element, for which the
>      * property has been cleared for.
> ```

remove last "for"

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 281 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp1WLjsUr9AJTzR38h:-Kkp1WLjsUr9AJTzR38i:b6ry8ks) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L281)):*
> ```Java
>             clonedElementCallback.accept(clonedElement);
>         }
>         draggedElement.getParentElement().appendChild(clonedElement);
> ```

could this have side effects?

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 341 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp41NsLc6xyDTdEOAy:-Kkp41NsLc6xyDTdEOAz:bb9end7) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L341)):*
> ```Java
>             clonedStyle.setProperty("transform", sb.toString());
>         }
>         // the x-offset should be taken into account when the drag image is
> ```

(why doesn't y matter?)

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 370 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp0m3CYDMSBCUnNxto:-Kkp0m3CYDMSBCUnNxtp:brkellu) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L370)):*
> ```Java
>         } catch (NumberFormatException nfe) {
>             Logger.getLogger(DragSourceExtensionConnector.class.getName())
>                     .warning(
> ```

(I guess info level might be sufficient)

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 456 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp0JHFwNXvusxVOkzV:-Kkp0JHFwNXvusxVOkzW:b3rhusy) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L456)):*
> ```Java
>             dragStartEvent.getDataTransfer().setDragImage(
>                     dragImage.getElement(),
>                     WidgetUtil.getRelativeX(draggedElement, dragStartEvent),
> ```

Does this work also if the translation changes later (e.g. animation, and this method is called at the beginning)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9409)*
<!-- Sent from Reviewable.io -->

Commit:
Fix Grid row DnD image for frozen cells, review comments

When there are frozen columns used, the image needs should not look weird because of the frozen column transitions.
The multiselection column is now not shown it is frozen.

pleku commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java, line 246 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp4BvkJvCDk-ExpggY:-KkpBCFpFi3aGjEfQcDW:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java#L246)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(could consolidate comment on two lines)
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java, line 475 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp45oK6Rz4e6Ya4YwO:-KkpBEA3jq78uF5AFRzw:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceConnector.java#L475)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(EOL missing)
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 237 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp1267nYQhnAYPxujk:-KkpBKS0Q-ecYpe7diU8:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L237)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

remove last "for"
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 281 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp1WLjsUr9AJTzR38h:-KkpBVtXwXXAT02kKIjC:b-auqkyp) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L281)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

could this have side effects?
</blockquote></details>

Moved z-index setting with comment to answer this, I don't know of any other side effects

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 341 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp41NsLc6xyDTdEOAy:-KkpE5L7gIdgRy1eMy5i:bcm4fot) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L341)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(why doesn't y matter?)
</blockquote></details>

Tried to anwer

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 370 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp0m3CYDMSBCUnNxto:-KkpCEHxPjmR794EKiho:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L370)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(I guess info level might be sufficient)
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 456 at r1](https://reviewable.io:443/reviews/vaadin/framework/9409#-Kkp0JHFwNXvusxVOkzV:-KkpCIxlx1gsmqnfQBmy:b-j2g2wg) ([raw file](https://github.com/vaadin/framework/blob/1f499d9e1da4888231b130a9763d7caecae30192/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L456)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Does this work also if the translation changes later (e.g. animation, and this method is called at the beginning)?
</blockquote></details>

Not sure if I understood the Q, but this uses the dragged element so whatever animations have been used are already applied (I think), the relativeness is calculated against the existing element not the cloned (which we hack here).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9409)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9409)*
<!-- Sent from Reviewable.io -->
